ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -1007,7 +1007,10 @@ void checkAndMaybeComplete() {
       if (progressTracker.isDone()) {
         if (progressTracker.hasSucceeded() && !retainChunkExceptionOnSuccess) {
           chunkException = null;
-        } else if (chunkOperationTracker.hasFailedOnNotFound()) {
+        } else if (chunkOperationTracker.maybeFailedDueToOfflineReplicas()) {
```

Comment:
AFAICT, maybeFailedDueToOfflineReplicas() can return true due to offline replicas on remote colos. I think we need to care only for origin colos.

Also, should we only consider offline replicas? What happens replicas are marked as online in the topology, yet they don't return a definite answer, e.g., requests time out

Reply:
Currently for TtlUpdate and Delete, we care about remote colo offline as well when routerUnavailableDueToOfflineReplicas is enabled.

We can either,
1. To be the same as delete and ttlupdate.
2. Differentiate the two cases,
    2.1 If delete and ttlupdate
    if (hasFailedOnCrossColoNotFound()
        && allReplicaCount - totalNotFoundCount >= replicaSuccessTarget
        && totalOfflineReplicaCount) return true;
    2.2 if read
    if (hasFailedOnCrossColoNotFound()
        && allReplicaCount - totalNotFoundCount >= replicaSuccessTarget
        && (originatingDcEligibleReplicaCount < originatingDcTotalReplicaCount - routerConfig.routerPutSuccessTarget + 1))
   

Timeout is also a valid error status.

Reply:
What do you mean by timeout being a valid error status? What would we do if we get one 404 and 8 timeouts?

Reply:
1. I refined the code. Implement maybeFailedDueToOfflineReplicas differently for GetBlob and TtlUpdate/Delete.

2. Time out is a specific error code RouterErrorCode.OperationTimedOut. It's different than the Not_Found code.
    When we translate router error code to Rest code, OperationTimedOut/AmbryUnavailable/RouterClosed all are translated to ServiceUnavailable;